### PR TITLE
ENH: Improve error message in multinomial

### LINKED ISSUE #8317: multinomial casts input to np.float64
`np.random.multinomial` seems to cast its second argument `pvals` to an array of dtype float64. Whilst this isn't an issue in and of itself, I've come across an interesting scenario where I have an array of dtype float32 whose sum is `0.99999994`, and when it gets cast to float64, its sum is now `1.0000000222053895`, which causes `np.random.multinomial` to raise `ValueError: sum(pvals[:-1]) > 1.0`. 

For a little context, I have an array of dtype float32 because it's come off the GPU, where single precision calculations are a great deal faster.

### PR DESCRIPTION
**Pull request by bashtage**
Improve error message when the sum of pvals is larger than 1
when the input data is an ndarray

closes #8317
xref #16732

### CODE REVIEW

**Comment by bashtage** [comment]
Supercedes #16732 with the simplest route which is an improved error in the edge case mentioned in the original issue.

File: numpy/random/_generator.pyx [hunk 1]
```diff
@@ -3749,7 +3749,17 @@ cdef class Generator:
...
+            msg = "sum(pvals[:-1]) > 1.0"
+            # When floating, but not float dtype, and close, improve the error
+            # 1.0001 works for float16 and float32
+            if (isinstance(pvals, np.ndarray) and
+                    pvals.dtype != float and
+                    np.issubdtype(pvals.dtype, np.floating) and
+                    pvals.sum() < 1.0001):
+                msg += (". pvals has been cast to double before checking "
+                        "the sum. Changes in precision when casting may "
+                        "produce violations even if pvals.sum() <= 1.")
```
Comment (eric-wieser) [inline numpy/random/_generator.pyx:3761]: What are your thoughts on including `astype` in the message?
```suggestion
            msg = "sum(pvals[:-1]) > 1.0"
            # When floating, but not float dtype, and close, improve the error
            # 1.0001 works for float16 and float32
            if (isinstance(pvals, np.ndarray) and
                    pvals.dtype != float and
                    np.issubdtype(pvals.dtype, np.floating) and
                    pvals.sum() < 1.0001):
                msg += ("pvals[:-1].astype(np.float64).sum() > 1.0. @
                        "Changes in precision when casting may "
                        "produce violations even if pvals.sum() <= 1.")
            else:
                msg = "sum(pvals[:-1]) > 1.0"
```

Comment (bashtage) [reply numpy/random/_generator.pyx]: I like its precision. I kept a bit more test than you.

File: numpy/random/_generator.pyx [hunk 2]
```diff
@@ -3749,7 +3749,19 @@ cdef class Generator:
...
+            # 1.0001 works for float16 and float32
+            if (isinstance(pvals, np.ndarray) and
+                    pvals.dtype != float and
+                    np.issubdtype(pvals.dtype, np.floating) and
+                    pvals.sum() < 1.0001):
+                msg = ("sum(pvals[:-1].astype(np.float64)) > 1.0. pvals are "
+                       "cast to 64-bit floating point values prior to "
+                       "checking the constraint. Changes in precision when "
+                       "casting may produce violations even if "
+                       "pvals[:-1].sum() <= 1.")
```
Comment (eric-wieser) [inline numpy/random/_generator.pyx:3763]: I think it's clearer to not replace the message.
```suggestion
            # When floating, but not float dtype, and close, improve the error
            # 1.0001 works for float16 and float32
            if (isinstance(pvals, np.ndarray) and
                    pvals.dtype != float and
                    np.issubdtype(pvals.dtype, np.floating) and
                    pvals.sum() < 1.0001):
                msg = ("sum(pvals[:-1].astype(np.float64)) > 1.0. pvals are "
                       "cast to 64-bit floating point values prior to "
                       "checking the constraint. Changes in precision when "
                       "casting may produce violations even if "
                       "pvals[:-1].sum() <= 1.")
            else:
                msg = "sum(pvals[:-1]) > 1.0"
```

File: numpy/random/_generator.pyx [hunk 3]
```diff
@@ -3749,7 +3749,20 @@ cdef class Generator:
         pix = <double*>np.PyArray_DATA(parr)
         check_array_constraint(parr, 'pvals', CONS_BOUNDED_0_1)
         if kahan_sum(pix, d-1) > (1.0 + 1e-12):
-            raise ValueError("sum(pvals[:-1]) > 1.0")
+            # When floating, but not float dtype, and close, improve the error
+            # 1.0001 works for float16 and float32
+            if (isinstance(pvals, np.ndarray) and
```
Comment (charris) [inline numpy/random/_generator.pyx:3754]: PEP8 advises that new code put the operators at the beginning of lines.

File: numpy/random/_generator.pyx [hunk 4]
```diff
@@ -3749,7 +3749,20 @@ cdef class Generator:
...
+                    pvals.dtype != float and
+                    np.issubdtype(pvals.dtype, np.floating) and
```
Comment (charris) [inline numpy/random/_generator.pyx:3756]: I'd reverse the order of these two lines.

File: numpy/random/_generator.pyx [hunk 5]
```diff
@@ -3749,7 +3749,20 @@ cdef class Generator:
...
+                    pvals.sum() < 1.0001):
+                msg = ("sum(pvals[:-1].astype(np.float64)) > 1.0. pvals are "
```
Comment (charris) [inline numpy/random/_generator.pyx:3758]: Maybe "The" before "pvals" so sentence starts with a capital.

File: numpy/random/_generator.pyx [hunk 6]
```diff
@@ -3749,7 +3749,20 @@ cdef class Generator:
...
+                       "cast to 64-bit floating point values prior to "
+                       "checking the constraint. Changes in precision when "
```
Comment (charris) [inline numpy/random/_generator.pyx:3760]: "constraint" -> "sum"
"Changes in precision" -> "Precision changes"
"produce violations" -> "cause problems"

File: numpy/random/_generator.pyx [hunk 7]
```diff
@@ -3749,7 +3749,20 @@ cdef class Generator:
...
+                       "casting may produce violations even if "
+                       "pvals[:-1].sum() <= 1.")
```
Comment (charris) [inline numpy/random/_generator.pyx:3762]: "The sum of the original pvals was valid."

File: numpy/random/_generator.pyx (hunk 3 above)
Comment (bashtage) [reply numpy/random/_generator.pyx]: Fixed.

File: numpy/random/_generator.pyx (hunk 5 above)
Comment (bashtage) [reply numpy/random/_generator.pyx]: Fixed.

File: numpy/random/_generator.pyx (hunk 4 above)
Comment (bashtage) [reply numpy/random/_generator.pyx]: Fixed.

File: numpy/random/_generator.pyx (hunk 7 above)
Comment (bashtage) [reply numpy/random/_generator.pyx]: Fixed. Thanks.

File: numpy/random/_generator.pyx (hunk 6 above)
Comment (bashtage) [reply numpy/random/_generator.pyx]: Fixed.

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
Thanks Kevin.

**Pull request merged**